### PR TITLE
Automated cherry pick of #60009: add node shutdown taint

### PR DESCRIPTION
**Pull request by lblackstone**
Cherry pick of #60009 on release-1.10.

#60009: Add taint on node shutdown to allow immediate volume detach

**Release note:**
```release-note
Add taint on node shutdown to allow immediate volume detach
```

### CODE REVIEW

**Comment by MaciekPytel** [comment]
@zetaab @lblackstone Can you edit description on #60009 and add a release note? In general changes that are cherry-picked should have release notes, so the users know what is changing in patch release.

**Comment by lblackstone** [comment]
@dims @caesarxuchao @aveshagarwal  
Still waiting for reviews on this cherry pick. Anything else I need to do here?

**Comment by waynr** [comment]
This looks good to me.

**Comment by jrperritt** [comment]
@dims @FengyunPan2 This may have gotten lost on the radar with other release work happening at the time, but we'd still like to see this get merged.

**Comment by MaciekPytel** [comment]
@lblackstone @zetaab Can we copy the release note from here to #60009? Release notes are actually generated from notes on parent PR, notes on cherry-picks are ignored.

**Comment by lblackstone** [comment]
Still need a couple approvals here to get this merged for 1.10.7

/assign @deads2k @timothysc

**Comment by lblackstone** [comment]
@MaciekPytel I don't have permission to change the release note on that PR. I haven't seen @zetaab respond to any pings since this was opened, so maybe another member could edit it?

**Comment by bsalamat** [comment]
/assign @k82cn 

Klaus, since you have reviewed the original PR, it makes more sense if you review this cherrypick as well.

**Comment by lblackstone** [comment]
Ping @deads2k -- still need an approval for this cherry pick

**Comment by mikedanese** [comment]
Is this a critical bug fix?

**Comment by lblackstone** [comment]
@mikedanese It's critical for OpenStack users at least. Without this patch, PVs can take on the order of 10 mins to detach from a node that failed (or was shutdown for some reason). Until it detaches, the associated workload will be unhealthy when it gets rescheduled.

**Comment by MaciekPytel** [comment]
/hold
Freezing branch for 1.10.7

**Comment by jsafrane** [comment]
release note edited

**Comment by therevoman** [comment]
@lblackstone and @mikedanese, currently working with a high profile client who is blocked by this issue because they've committed to Kubernetes on OpenStack.

**Comment by nikhita** [comment]
Closing and reopening to retrigger prow to remove the `do-not-merge/cherry-pick-not-approved` label.

/assign

**Comment by nikhita** [comment]
/unassign

The labels are fixed now. :+1:

**Comment by lblackstone** [comment]
Ping @mikedanese @deads2k -- Still need an approval from `pkg/controller/OWNERS`

**Comment by gnufied** [comment]
@zetaab @jingxu97 correct me if I am wrong but this PR in itself will not cause immediate detach of volumes that we were expecting. We will also need something like https://github.com/kubernetes/kubernetes/pull/66213 to trigger actual detach. If that is the case - what is the point of backporting this PR to 1.10?

cc @jsafrane @deads2k

**Comment by yastij** [comment]
+1 @gnufied

**Comment by zetaab** [comment]
@gnufied this does not implement immediate detach. Actually two things is missing before we could detach anything. 1) all cloud providers is missing `InstanceShutdownByProviderID` function. Like this it does not do anything (it does not add taint label at all). 2) volume controllers does not have anything implemented.

tl;dr this should not affect anything before those two things are implemented

**Comment by gnufied** [comment]
@zetaab right, so we should not backport this IMO. Because the logic to perform actual detach has been feature gated and we can't backport that.

**Comment by MaciekPytel** [comment]
/hold 
For the ongoing discussion.

**Comment by zetaab** [comment]
@gnufied yea I agree because this does not do anything if backported

**Comment by lblackstone** [comment]
It seems that I misunderstood the scope of this change, and the important functionality (detaching PVs on a shutdown node) isn't actually implemented yet. In light of that information, I don't think the cherry-pick makes sense for 1.10.x. Closing it now, and sorry for the confusion.
